Sources/PointFree/Auth.swift

```diff
@@ -0,0 +1,123 @@
+import Either
+import Foundation
+import Html
+import HttpPipeline
+import HttpPipelineHtmlSupport
+import Optics
+import Prelude
+
+let secretHomeResponse: (Conn<StatusLineOpen, Prelude.Unit>) -> IO<Conn<ResponseEnded, Data?>> =
+  writeStatus(.ok)
+    >>> readGitHubSessionCookieMiddleware
+    >>> respond(secretHomeView)
+    >>> pure
+
+let githubCallbackResponse =
+  authTokenMiddleware
+    >-> redirect(to: link(to: .secretHome), headersMiddleware: writeGitHubSessionCookieMiddleware)
+    >>> pure
+
+let loginResponse: (Conn<StatusLineOpen, Prelude.Unit>) -> IO<Conn<ResponseEnded, Data?>> =
+  redirect(to: githubAuthorizationUrl)
+    >>> pure
+
+let logoutResponse: (Conn<StatusLineOpen, Prelude.Unit>) -> IO<Conn<ResponseEnded, Data?>> =
+  redirect(
+    to: link(to: .secretHome),
+    headersMiddleware: writeHeader(.clearCookie(key: "github_session"))
+    )
+    >>> pure
+
+private let secretHomeView = View<Either<Prelude.Unit, GitHubUserEnvelope>> { data in
+  [
+    p(["welcome home"]),
+
+    p([
+      text(
+        data.right.map { "You are logged in as \($0.gitHubUser.name)" }
+          ?? "You are not logged in"
+      )
+      ]),
+
+    p([
+       data.isRight
+        ? a([href(link(to: .logout))], ["Log out"])
+        : a([href(link(to: .login))], ["Log in"])
+      ])
+    ]
+}
+
+// todo: move to swift-web
+extension URLRequest {
+  public var cookies: [String: String] {
+    let pairs = (self.allHTTPHeaderFields?["Cookie"] ?? "")
+      .components(separatedBy: "; ")
+      .map {
+        $0.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
+          .map(String.init)
+      }
+      .flatMap { pure(createTuple) <*> $0.first <*> $0.last }
+    return .init(uniqueKeysWithValues: pairs)
+  }
+}
+
+// todo: move to prelude
+private func createTuple<A, B>(_ a: A) -> (B) -> (A, B) {
+  return { b in (a, b) }
+}
+
+private func readGitHubSessionCookieMiddleware(
+  _ conn: Conn<HeadersOpen, Prelude.Unit>
+  )
+  -> Conn<HeadersOpen, Either<Prelude.Unit, GitHubUserEnvelope>> {
+
+    return conn.map(
+      const(
+        conn.request.cookies["github_session"]
+          .flatMap { ResponseHeader.verifiedValue(signedCookieValue: $0, secret: EnvVars.appSecret) }
+          .map(Either.right)
+          ?? Either.left(unit)
+      )
+    )
+}
+
+private func writeGitHubSessionCookieMiddleware(
+  _ conn: Conn<HeadersOpen, Either<Prelude.Unit, GitHubUserEnvelope>>
+  )
+  -> Conn<HeadersOpen, Either<Prelude.Unit, GitHubUserEnvelope>> {
+
+    switch conn.data {
+    case .left:
+      return conn
+    case let .right(envelope):
+      return conn |> writeHeaders(
+        [
+          ResponseHeader.setSignedCookie(
+            key: "github_session",
+            value: envelope,
+            options: [],
+            secret: EnvVars.appSecret,
+            encrypt: true
+          )
+          ] |> catOptionals
```

Comment:
not sure how I feel about `ResponseHeader.setSignedCookie` returning an optional header :/ kinda weird to have to `catOptionals` everything

Reply:
Could we handle the failure explicitly?

Reply:
yeah we could make this a middleware transformer so that we could do a redirect to an error / show a flash message. something to try out soon...